File: examples/algorithm-deepening/pattern-lab-v06/greedy_interval_trace.cpp
#include <algorithm>
#include <iostream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <vector>

struct Interval { std::string label; int start; int end; };

void validate(const std::vector<Interval>& intervals) {
  for (const auto& item : intervals) {
    if (item.end < item.start) throw std::invalid_argument("invalid interval");
  }
}

std::vector<Interval> select(std::vector<Interval> intervals, bool by_finish) {
  validate(intervals);
  std::sort(intervals.begin(), intervals.end(), [by_finish](const auto& a, const auto& b) {
    if (by_finish) return std::tie(a.end, a.start, a.label) < std::tie(b.end, b.start, b.label);
    return std::tie(a.start, a.end, a.label) < std::tie(b.start, b.end, b.label);
  });
  std::vector<Interval> result;
  for (const auto& item : intervals) {
    if (result.empty() || item.start >= result.back().end) result.push_back(item);
  }
  return result;
}

std::vector<Interval> sample() {
  return {{"A",1,4},{"B",3,5},{"C",0,6},{"D",5,7},{"E",3,9},{"F",5,9},
          {"G",6,10},{"H",8,11},{"I",8,12},{"J",2,14},{"K",12,16}};
}

void print_labels(const std::vector<Interval>& values) {
  for (std::size_t index=0; index<values.size(); ++index) {
    if (index>0) std::cout << ',';
    std::cout << values[index].label;
  }
}

int main() {
  auto intervals=sample();
  auto ordered=intervals;
  std::sort(ordered.begin(), ordered.end(), [](const auto& a,const auto& b){
    return std::tie(a.end,a.start,a.label)<std::tie(b.end,b.start,b.label);
  });
  const auto greedy=select(intervals,true);
  const auto wrong=select(intervals,false);
  std::cout<<"intervals=A[1,4),B[3,5),C[0,6),D[5,7),E[3,9),F[5,9),G[6,10),H[8,11),I[8,12),J[2,14),K[12,16)\n";
  std::cout<<"order_by_finish=";print_labels(ordered);std::cout<<'\n';
  std::cout<<"select=";print_labels(greedy);std::cout<<" count="<<greedy.size()<<'\n';
  std::cout<<"earliest_start_select=";print_labels(wrong);std::cout<<" count="<<wrong.size()<<'\n';
  std::cout<<"exchange=finish-no-later-preserves-room\n";
  std::cout<<"invariant=selected-intervals-nonoverlap\n";
}

File: examples/algorithm-deepening/pattern-lab-v06/greedy_interval_trace.py
from __future__ import annotations

from dataclasses import dataclass
from itertools import combinations
from typing import Sequence


@dataclass(frozen=True)
class Interval:
    label: str
    start: int
    end: int


def _validate(intervals: Sequence[Interval]) -> None:
    if any(item.end < item.start for item in intervals):
        raise ValueError("interval end must not precede start")


def earliest_finish_schedule(intervals: Sequence[Interval]) -> tuple[Interval, ...]:
    _validate(intervals)
    ordered = sorted(intervals, key=lambda item: (item.end, item.start, item.label))
    selected: list[Interval] = []
    last_end: int | None = None
    for item in ordered:
        if last_end is None or item.start >= last_end:
            selected.append(item)
            last_end = item.end
    return tuple(selected)


def earliest_start_schedule(intervals: Sequence[Interval]) -> tuple[Interval, ...]:
    _validate(intervals)
    ordered = sorted(intervals, key=lambda item: (item.start, item.end, item.label))
    selected: list[Interval] = []
    last_end: int | None = None
    for item in ordered:
        if last_end is None or item.start >= last_end:
            selected.append(item)
            last_end = item.end
    return tuple(selected)


def maximum_count_bruteforce(intervals: Sequence[Interval]) -> int:
    _validate(intervals)
    best = 0
    for size in range(len(intervals) + 1):
        for subset in combinations(intervals, size):
            ordered = sorted(subset, key=lambda item: (item.start, item.end, item.label))
            if all(ordered[index].start >= ordered[index - 1].end for index in range(1, len(ordered))):
                best = max(best, size)
    return best


def sample_intervals() -> list[Interval]:
    return [
        Interval("A", 1, 4), Interval("B", 3, 5), Interval("C", 0, 6),
        Interval("D", 5, 7), Interval("E", 3, 9), Interval("F", 5, 9),
        Interval("G", 6, 10), Interval("H", 8, 11), Interval("I", 8, 12),
        Interval("J", 2, 14), Interval("K", 12, 16),
    ]


def fixed_report() -> str:
    intervals = sample_intervals()
    ordered = sorted(intervals, key=lambda item: (item.end, item.start, item.label))
    greedy = earliest_finish_schedule(intervals)
    wrong = earliest_start_schedule(intervals)
    return "\n".join(
        [
            "intervals=A[1,4),B[3,5),C[0,6),D[5,7),E[3,9),F[5,9),G[6,10),H[8,11),I[8,12),J[2,14),K[12,16)",
            f"order_by_finish={','.join(item.label for item in ordered)}",
            f"select={','.join(item.label for item in greedy)} count={len(greedy)}",
            f"earliest_start_select={','.join(item.label for item in wrong)} count={len(wrong)}",
            "exchange=finish-no-later-preserves-room",
            "invariant=selected-intervals-nonoverlap",
        ]
    )


if __name__ == "__main__":
    print(fixed_report())
